File: tesseract-decoder/src/py/common_test.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http:#www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import pytest
import stim
import tesseract_decoder


def get_set_bits(n):
    """
    Converts an observable bitmask (integer) into a list of observable indices.

    Args:
        n (int): The integer representing the observable bitmask.

    Returns:
        list[int]: A list containing the indices of the set bits (observable IDs)
    """
    bits = []
    i = 0

    while n > 0:
        if n & 1:
            bits.append(i)
        n >>= 1
        i += 1
    return bits


def test_error_from_direct_constructor():
    # Test the new constructor with likelihood_cost, detectors, and observables
    likelihood_cost = 1.945910
    detectors = [1, 2]
    observables = get_set_bits(4324)
    error = tesseract_decoder.common.Error(likelihood_cost, detectors, observables)

    assert error.likelihood_cost == pytest.approx(likelihood_cost)
    assert error.symptom.detectors == detectors
    assert error.symptom.observables == observables


def test_error_str():
    # Test the new constructor with likelihood_cost, detectors, and observables
    likelihood_cost = 5.5
    detectors = [1, 2]
    observables = [5, 10]
    error = tesseract_decoder.common.Error(likelihood_cost, detectors, observables)

    assert (
        str(error)
        == "Error{cost=5.500000, symptom=Symptom{detectors=[1 2], observables=[5 10]}}"
    )


def test_as_dem_instruction_targets():
    s = tesseract_decoder.common.Symptom([1, 2], get_set_bits(4324))
    dits = s.as_dem_instruction_targets()
    assert dits == [
        stim.DemTarget("D1"),
        stim.DemTarget("D2"),
        stim.DemTarget("L2"),
        stim.DemTarget("L5"),
        stim.DemTarget("L6"),
        stim.DemTarget("L7"),
        stim.DemTarget("L12"),
    ]


def test_error_from_dem_instruction():
    di = stim.DemInstruction("error", [0.125], [stim.target_logical_observable_id(3)])
    error = tesseract_decoder.common.Error(di)
    assert (
        str(error)
        == "Error{cost=1.945910, symptom=Symptom{detectors=[], observables=[3]}}"
    )


def test_error_get_set_probability():
    error = tesseract_decoder.common.Error()
    probability = 0.125
    expected_cost = 1.9459101490553132

    error.set_with_probability(probability)
    assert error.likelihood_cost == pytest.approx(expected_cost)
    assert error.get_probability() == pytest.approx(probability)

    probability = 0.5
    expected_cost = 0.0

    error.set_with_probability(probability)
    assert error.likelihood_cost == pytest.approx(expected_cost)
    assert error.get_probability() == pytest.approx(probability)


def test_error_set_with_probability_invalid_input():
    error = tesseract_decoder.common.Error()

    with pytest.raises(ValueError):
        error.set_with_probability(0.0)

    with pytest.raises(ValueError):
        error.set_with_probability(1.0)

    with pytest.raises(ValueError):
        error.set_with_probability(-0.1)

    with pytest.raises(ValueError):
        error.set_with_probability(1.1)


def test_merge_indistinguishable_errors():
    dem = stim.DetectorErrorModel()
    assert isinstance(
        tesseract_decoder.common.merge_indistinguishable_errors(dem),
        stim.DetectorErrorModel,
    )


def test_remove_zero_probability_errors():
    dem = stim.DetectorErrorModel()
    assert isinstance(
        tesseract_decoder.common.remove_zero_probability_errors(dem),
        stim.DetectorErrorModel,
    )


def test_dem_from_counts():
    dem = stim.DetectorErrorModel()
    assert isinstance(
        tesseract_decoder.common.dem_from_counts(dem, [], 3), stim.DetectorErrorModel
    )


if __name__ == "__main__":
    raise SystemExit(pytest.main([__file__]))

File: deps/tesseract/src/common.test.cc
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "common.h"

#include <limits>

#include "gtest/gtest.h"
#include "stim.h"

namespace tesseract_decoder {
namespace {

stim::Circuit source_circuit() {
  return stim::Circuit(R"CIRCUIT(
    M 0 1
    DETECTOR rec[-1]
    DETECTOR rec[-2]
    OBSERVABLE_INCLUDE(0) rec[-1]
  )CIRCUIT");
}

TEST(common, ErrorsStructFromDemInstruction) {
  // Test a pathological DEM error instruction
  stim::DetectorErrorModel dem("error(0.1) D0 ^  D0 D1  L0 L1 L1");
  stim::DemInstruction instruction = dem.instructions.at(0);
  common::Error ES(instruction);
  EXPECT_EQ(ES.symptom.detectors, std::vector<int>{1});
  EXPECT_EQ(ES.symptom.observables, std::vector<int>{0});
}

TEST(common, ShotDetectorCountAllowsSourceAlignedDemAugmentation) {
  const stim::Circuit circuit = source_circuit();
  const stim::DetectorErrorModel augmented_dem(R"DEM(
    error(0.1) D0 L0
    detector D2
  )DEM");

  EXPECT_EQ(common::shot_detector_count(circuit, augmented_dem), 2);
}

TEST(common, ShotDetectorCountRejectsMissingCircuitDetectors) {
  const stim::Circuit circuit = source_circuit();
  const stim::DetectorErrorModel too_small_dem("error(0.1) D0 L0");

  EXPECT_THROW(common::shot_detector_count(circuit, too_small_dem), std::invalid_argument);
}

TEST(common, ShotDetectorCountRejectsDifferentObservableCounts) {
  const stim::Circuit circuit = source_circuit();
  const stim::DetectorErrorModel wrong_observable_count_dem(R"DEM(
    error(0.1) D0 L1
    detector D2
  )DEM");

  EXPECT_THROW(common::shot_detector_count(circuit, wrong_observable_count_dem),
               std::invalid_argument);
}

TEST(common, DemFromCountsHandlesZeroProbabilityErrors) {
  stim::DetectorErrorModel dem(R"DEM(
        error(0.1) D0
        error(0) D1
        error(0.2) D2
        detector(0, 0, 0) D0
        detector(0, 0, 0) D1
        detector(0, 0, 0) D2
      )DEM");

  std::vector<size_t> counts{1, 7, 4};
  size_t num_shots = 10;
  stim::DetectorErrorModel out_dem = common::dem_from_counts(dem, counts, num_shots);

  auto flat = out_dem.flattened();
  ASSERT_EQ(out_dem.count_errors(), 3);
  ASSERT_GE(flat.instructions.size(), 3);

  EXPECT_EQ(flat.instructions[0].type, stim::DemInstructionType::DEM_ERROR);
  EXPECT_NEAR(flat.instructions[0].arg_data[0], 0.1, 1e-9);
  EXPECT_EQ(flat.instructions[1].type, stim::DemInstructionType::DEM_ERROR);
  EXPECT_NEAR(flat.instructions[1].arg_data[0], 0.7, 1e-9);
  EXPECT_EQ(flat.instructions[2].type, stim::DemInstructionType::DEM_ERROR);
  EXPECT_NEAR(flat.instructions[2].arg_data[0], 0.4, 1e-9);

  std::vector<size_t> error_index_map;
  stim::DetectorErrorModel cleaned = common::remove_zero_probability_errors(dem, error_index_map);
  stim::DetectorErrorModel out_dem_cleaned =
      common::dem_from_counts(cleaned, std::vector<size_t>{1, 4}, num_shots);

  auto flat_cleaned = out_dem_cleaned.flattened();
  ASSERT_EQ(out_dem_cleaned.count_errors(), 2);
  ASSERT_GE(flat_cleaned.instructions.size(), 2);

  EXPECT_EQ(flat_cleaned.instructions[0].type, stim::DemInstructionType::DEM_ERROR);
  EXPECT_NEAR(flat_cleaned.instructions[0].arg_data[0], 0.1, 1e-9);
  ASSERT_EQ(flat_cleaned.instructions[1].type, stim::DemInstructionType::DEM_ERROR);
  EXPECT_NEAR(flat_cleaned.instructions[1].arg_data[0], 0.4, 1e-9);
}

TEST(common, DemFromCountsSimpleTwoErrors) {
  stim::DetectorErrorModel dem(R"DEM(
        error(0.25) D0
        error(0.35) D1
        detector(0, 0, 0) D0
        detector(0, 0, 0) D1
      )DEM");

  std::vector<size_t> counts{5, 7};
  size_t num_shots = 20;
  stim::DetectorErrorModel out_dem = common::dem_from_counts(dem, counts, num_shots);

  auto flat = out_dem.flattened();
  ASSERT_EQ(out_dem.count_errors(), 2);

  ASSERT_GE(flat.instructions.size(), 2);
  EXPECT_EQ(flat.instructions[0].type, stim::DemInstructionType::DEM_ERROR);
  EXPECT_NEAR(flat.instructions[0].arg_data[0], 0.25, 1e-9);
  EXPECT_EQ(flat.instructions[1].type, stim::DemInstructionType::DEM_ERROR);
  EXPECT_NEAR(flat.instructions[1].arg_data[0], 0.35, 1e-9);
}

TEST(common, RemoveZeroProbabilityErrors) {
  stim::DetectorErrorModel dem(R"DEM(
        error(0.1) D0
        error(0) D1
        error(0.2) D2
        detector(0, 0, 0) D0
        detector(0, 0, 0) D1
        detector(0, 0, 0) D2
      )DEM");

  std::vector<size_t> error_index_map;
  stim::DetectorErrorModel cleaned = common::remove_zero_probability_errors(dem, error_index_map);

  EXPECT_EQ(cleaned.count_errors(), 2);
  auto flat = cleaned.flattened();
  ASSERT_EQ(flat.instructions[0].type, stim::DemInstructionType::DEM_ERROR);
  EXPECT_NEAR(flat.instructions[0].arg_data[0], 0.1, 1e-9);
  ASSERT_EQ(flat.instructions[1].type, stim::DemInstructionType::DEM_ERROR);
  EXPECT_NEAR(flat.instructions[1].arg_data[0], 0.2, 1e-9);
}

TEST(common, RemoveZeroProbabilityErrorsPreservesIndexSpaces) {
  stim::DetectorErrorModel dem("error(0) D2 L1");

  std::vector<size_t> error_index_map;
  stim::DetectorErrorModel cleaned = common::remove_zero_probability_errors(dem, error_index_map);

  EXPECT_EQ(cleaned.count_errors(), 0);
  EXPECT_EQ(cleaned.count_detectors(), 3);
  EXPECT_EQ(cleaned.count_observables(), 2);
  EXPECT_EQ(error_index_map, (std::vector<size_t>{std::numeric_limits<size_t>::max()}));
}

TEST(common, MergeIndistinguishableErrorsPreservesIndexSpaces) {
  stim::DetectorErrorModel dem("error(0.1) D2 D2 L1 L1");

  std::vector<size_t> error_index_map;
  stim::DetectorErrorModel merged = common::merge_indistinguishable_errors(dem, error_index_map);

  EXPECT_EQ(merged.count_errors(), 1);
  EXPECT_EQ(merged.count_detectors(), 3);
  EXPECT_EQ(merged.count_observables(), 2);
}

// Helper function to compare the two methods.
void assert_merged_probabilities_are_equal(double p1, double p2) {
  // Merge probabilities using the exclusive OR formula.
  double merged_p_direct = p1 + p2 - 2 * p1 * p2;

  // Convert to likelihood costs, merge, and then convert back.
  double cost1 = std::log(p1 / (1 - p1));
  double cost2 = std::log(p2 / (1 - p2));
  double merged_cost = common::merge_weights(cost1, cost2);
  double merged_p_via_costs = 1 / (1 + std::exp(merged_cost));

  // The two methods should produce nearly same results.
  ASSERT_NEAR(merged_p_direct, merged_p_via_costs, 1e-12);
}

TEST(CommonTest, merge_weights_is_equivalent_to_probability_xor) {
  // Test with small probabilities
  assert_merged_probabilities_are_equal(0.001, 0.002);

  // Test with larger probabilities
  assert_merged_probabilities_are_equal(0.1, 0.25);

  // Test with a mix of small and large probabilities
  assert_merged_probabilities_are_equal(0.05, 0.8);

  // Test with a probability close to 0.5, where the formula is sensitive
  assert_merged_probabilities_are_equal(0.49, 0.51);

  // Test with identical probabilities
  assert_merged_probabilities_are_equal(0.01, 0.01);
}

// Helper function to create a simple DEM with two identical errors.
stim::DetectorErrorModel create_dem_with_two_errors(double p1, double p2) {
  stim::DetectorErrorModel dem;
  std::vector<stim::DemTarget> targets = {stim::DemTarget::relative_detector_id(0)};

  dem.append_error_instruction(p1, targets, "");
  dem.append_error_instruction(p2, targets, "");
  return dem;
}

// Function to get the probability from a merged DEM.
double get_merged_probability(const stim::DetectorErrorModel& merged_dem) {
  if (merged_dem.instructions.size() != 1 ||
      merged_dem.instructions[0].type != stim::DemInstructionType::DEM_ERROR) {
    throw std::runtime_error("Expected a single DEM_ERROR instruction in the merged model.");
  }
  return merged_dem.instructions[0].arg_data[0];
}

TEST(CommonTest, merge_indistinguishable_errors_two_errors) {
  // Case 1: Both probabilities are low.
  double p1 = 0.1;
  double p2 = 0.2;
  double expected_merged_p = p1 * (1 - p2) + p2 * (1 - p1);
  auto dem1 = create_dem_with_two_errors(p1, p2);
  std::vector<size_t> error_index_map;
  auto merged_dem1 = common::merge_indistinguishable_errors(dem1, error_index_map);
  ASSERT_NEAR(get_merged_probability(merged_dem1), expected_merged_p, 1e-9);

  // Case 2: One low, one high probability.
  p1 = 0.1;
  p2 = 0.8;
  expected_merged_p = p1 * (1 - p2) + p2 * (1 - p1);
  auto dem2 = create_dem_with_two_errors(p1, p2);
  auto merged_dem2 = common::merge_indistinguishable_errors(dem2, error_index_map);
  ASSERT_NEAR(get_merged_probability(merged_dem2), expected_merged_p, 1e-9);

  // Case 3: One high, one low probability.
  p1 = 0.8;
  p2 = 0.1;
  expected_merged_p = p1 * (1 - p2) + p2 * (1 - p1);
  auto dem3 = create_dem_with_two_errors(p1, p2);
  auto merged_dem3 = common::merge_indistinguishable_errors(dem3, error_index_map);
  ASSERT_NEAR(get_merged_probability(merged_dem3), expected_merged_p, 1e-9);

  // Case 4: Both probabilities are high.
  p1 = 0.8;
  p2 = 0.9;
  expected_merged_p = p1 * (1 - p2) + p2 * (1 - p1);
  auto dem4 = create_dem_with_two_errors(p1, p2);
  auto merged_dem4 = common::merge_indistinguishable_errors(dem4, error_index_map);
  ASSERT_NEAR(get_merged_probability(merged_dem4), expected_merged_p, 1e-9);
}

}  // namespace
}  // namespace tesseract_decoder
